datalad/core/distributed/clone.py

```diff
@@ -636,3 +691,101 @@ def _get_installationpath_from_url(url):
     if path.endswith('.git'):
         path = path[:-4]
     return path
+
+
+def decode_source_spec(spec, cfg=None):
+    """Decode information from a clone source specification
+
+    Parameters
+    ----------
+    spec : str
+      Any supported clone source specification
+    cfg : ConfigManager, optional
+      Configuration will be queried from the instance (i.e. from a particular
+      dataset). If None is given, the global DataLad configuration will be
+      queried.
+
+    Returns
+    -------
+    dict
+      The value of each decoded property is stored under its own key in this
+      dict. By default the following keys are return: 'type', a specification
+      type label {'giturl', 'dataladri', 'ria'}; 'source' the original
+      source specification; 'giturl' a URL for the source that is a suitable
+      source argument for git-clone; 'version' a version-identifer, if present
+      (None else); 'default_destpath' a relative path that that can be used as
+      a clone destination.
+    """
+    if cfg is None:
+        from datalad import cfg
+    # standard property dict composition
+    props = dict(
+        source=spec,
+        version=None,
+    )
+    # common starting point is a RI instance, support for accepting an RI
+    # instance is kept for backward-compatibility reasons
+    source_ri = RI(spec) if not isinstance(spec, RI) else spec
+
+    # scenario switch, each case must set 'giturl' at the very minimum
+    if isinstance(source_ri, DataLadRI):
+        # we have got our DataLadRI as the source, so expand it
+        props['type'] = 'dataladri'
+        props['giturl'] = source_ri.as_git_url()
+    elif isinstance(source_ri, URL) and source_ri.scheme.startswith('ria+'):
+        # parse a RIA URI
+        # check for store related configuration for SSH access
+        if source_ri.scheme == 'ria+ssh':
+            # we support the scenario were hostname is just a label for a config
+            # not an actual hostname
+            hostname = cfg.get(
+                'annex.ria-remote.{}.ssh-host'.format(source_ri.hostname),
+                source_ri.hostname)
+            basepath = cfg.get(
+                'annex.ria-remote.{}.base-path'.format(source_ri.hostname),
```

Comment:
If I understand the role of these options, I think `url.<base>.insteadOf` could also be used to handle the URL rewrites, in which case pointing users to a more general existing solution might be better than adding more options on our side.

Reply:
The original intend for those options is about the special remote itself, which doesn't pull its configuration from `git-annex:remote.log` but from those, allowing to reconfigure the mapping of an (autoenabled) special remote to the actual store being used by it depending on config. This was intended to easily switch the store depending on where a clone is (infrastructure reasons) based on a system config the user doesn't have to care about.

However, as the RIA approach evolves, we do indeed want to switch to something like you propose here, since we now can have all required info in that kind of URL. But this needs to be in sync with git-annex-ria-remote itself, so I'd postpone that change wrt this PR.

Reply:
I like the `insteadOf` idea. The consequence for this PR would be to simply remove the config query, and let lower-level code handle any conversion. I will do just that.

Reply:
Done.